contracts/Pool.sol

```diff
@@ -1,31 +1,314 @@
+    /// frozen between being taken from senders' balances and being collectable by the receiver.
+    /// High values make collecting cheaper by making it process less cycles for a given time range.
+    constructor(uint64 _cycleBlocks) public {
+        cycleBlocks = _cycleBlocks;
     }
 
-    function topUpSender(address sender) public payable {
-        senders[sender].balance += msg.value;
+    /// @notice Returns amount of received funds available for collection
+    /// by the sender of the message
+    /// @return collected The available amount
+    function collectable() public view returns (uint256) {
+        Receiver storage receiver = receivers[msg.sender];
+        uint64 collectedCycle = receiver.nextCollectedCycle;
+        if (collectedCycle == 0) return 0;
+        uint256 collected = 0;
+        int256 lastFundsPerCycle = receiver.lastFundsPerCycle;
+        uint256 currFinishedCycle = block.number / cycleBlocks;
+        for (; collectedCycle <= currFinishedCycle; collectedCycle++) {
+            lastFundsPerCycle += receiver.amtDeltas[collectedCycle];
+            collected += uint256(lastFundsPerCycle);
+        }
+        return collected;
     }
 
-    function withdraw(uint256 value) public {
-        withdrawTo(value, msg.sender);
+    /// @notice Collects all received funds available for collection
+    /// by a sender of the message and sends them to that sender
+    function collect() public {
+        Receiver storage receiver = receivers[msg.sender];
+        uint64 collectedCycle = receiver.nextCollectedCycle;
+        if (collectedCycle == 0) return;
+        uint256 currFinishedCycle = block.number / cycleBlocks;
+        if (collectedCycle > currFinishedCycle) return;
+        uint256 collected = 0;
+        int256 lastFundsPerCycle = receiver.lastFundsPerCycle;
+        for (; collectedCycle <= currFinishedCycle; collectedCycle++) {
+            int256 delta = receiver.amtDeltas[collectedCycle];
+            if (delta != 0) {
+                delete receiver.amtDeltas[collectedCycle];
+                lastFundsPerCycle += delta;
+            }
+            collected += uint256(lastFundsPerCycle);
+        }
+        receiver.lastFundsPerCycle = uint192(lastFundsPerCycle);
+        receiver.nextCollectedCycle = collectedCycle;
+        if (collected > 0) msg.sender.transfer(collected);
+    }
+
+    /// @notice Tops up the sender balance of a sender of the message with the amount in the message
+    function topUp() public payable suspendPayments {
+        senders[msg.sender].startBalance += uint192(msg.value);
+    }
+
+    /// @notice Returns amount of unsent funds available for withdrawal by the sender of the message
+    /// @return balance The available balance
+    function withdrawable() public view returns (uint256) {
+        Sender storage sender = senders[msg.sender];
+        uint256 amtPerWeight = sender.amtPerBlock / sender.weightSum;
```

Comment:
We need to guard this against the case where `sender.weightSum == 0`. Otherwise this will error if, for example, the caller does not have a pool yet. Same for `amtPerBlock` below.

Reply:
Great catch, fixed :facepalm:

Reply:
Could you also fix it for `amtPerBlock` below?

Reply:
Thanks for double checking, fixed, this time for real. I've also found a nasty bug right next to it.